### PR TITLE
#167728020 Password reset via email

### PR DESCRIPTION
**Pull request by dieudonneAwa**
**What does this PR do?**
- Resets user password via email

**Description of Task to be completed?**
Have the following Endpoints working
- POST `/api/v1/auth/forgotPassword`
- PUT `/api/v1//auth/resetPassword/:uuid/:token`

**How should this be manually tested?**
After cloning the repo,
- Create a SendGrid [account.](https://sendgrid.com/docs/)
- Create an API Key.
- Create an `.env` file and set `SENDGRID_API_KEY="The API key you created"`
- In that same file, make sure your database url is set as well
- Run migrations
- Run all seeds
- And it will be good to go

**Any background context you want to provide?**
To know more about how to send emails using sendGrid, go to the [Getting started with sendGrid API](https://sendgrid.com/docs/for-developers/sending-email/api-getting-started/)

**What are the relevant stories?**
- [#167728020](https://www.pivotaltracker.com/n/projects/2381984/stories/167728020)

**Screenshots (if appropriate)**
Screenshot (87)

**Questions:**
- None

### CODE REVIEW

**Comment by Abobos** [comment]
Well-done.. I think your endpoint  has to be something like api/v1/auth/forgotPassword and api/v1/auth/resetPassword since we are dealing with user authentication...... And also the reaetPassword should allow only accept token as it's query parameter..we don't need the user ID,  since it will be encoded in the token..

File: src/controllers/AuthController.js [hunk 1]
```diff
@@ -41,4 +48,65 @@ export default class AuthController {
...
+  * 
+  * @memberof UserController
+  */
+  static async sendResetLink(req, res) {
+    const { email } = req.body;
+    if (!inValidEmail(email)) {
+      const user = await models.User.findOne({ where: { email } });
+
+      if (!user) {
+        return sendErrorResponse(res, 404, `No user found with email address of ${email}`);
```
Comment (Abobos) [inline src/controllers/AuthController.js]: Why don't we use "Invalid credentials "instead of the long message we are returning to the user

**Comment by dieudonneAwa** [comment]
> Well-done.. I think your endpoint has to be something like api/v1/auth/forgotPassword and api/v1/auth/resetPassword since we are dealing with user authentication...... And also the reaetPassword should allow only accept token as it's query parameter..we don't need the user ID, since it will be encode in the token..

Thanks for your review. 
I will change the url from simply `/forgot` to ` /forgoPassword`, I thought of doing so but forgot to. 
However, adding the ID in the request parameter helps reduce the code size, which will also reduce the stress of using the verifyToken method more than once. And besides, only the user who is resetting his password has access to that particular endpoint, which means it is secured enough. I hope you get my point?

File: src/controllers/AuthController.js [hunk 2]
```diff
@@ -41,4 +48,65 @@ export default class AuthController {
...
+  * @param {Object} res - Response object
+  * 
+  * @returns {Object} object containing user data which will be embedded in link sent to user
+  * 
+  * @memberof UserController
+  */
+  static async sendResetLink(req, res) {
+    const { email } = req.body;
+    if (!inValidEmail(email)) {
+      const user = await models.User.findOne({ where: { email } });
```
Comment (justinefe) [inline src/controllers/AuthController.js]: I think the findOne should be in the userRepository

Comment (dieudonneAwa) [reply src/controllers/AuthController.js]: Thanks for the reminder

**Review by Kingobaino — APPROVED** [review]

**Review by Wokoro — APPROVED** [review]
LGTM

**Review by richienabuk — APPROVED** [review]

**Review by Abobos — APPROVED** [review]

**Pull request merged**